tcclient/creds.go

```diff
@@ -7,16 +7,75 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"net/http"
 	"strconv"
 	"strings"
 	"time"
 
 	"github.com/taskcluster/jsonschema2go/text"
 	"github.com/taskcluster/slugid-go/slugid"
+	"github.com/tent/hawk-go"
 )
 
-type Credentials struct {
-	ClientId    string `json:"clientId"`
+type Credentials interface {
+	fmt.Stringer
+	ConfigureAuth(httpRequest *http.Request) error
+	getExtField() (string, error)
+	HawkCredentials() *hawk.Credentials
```

Comment:
Well, if hide the fields in the struct then you can't just use the native
json.Marshal.... That's why I suggest we implement the interface. That way
credentials can be immutable, but still serializable to json...
Den 22. mar. 2016 1.08 AM skrev "Pete Moore" <notifications@github.com>:

> In tcclient/creds.go
> <https://github.com/taskcluster/taskcluster-client-go/pull/13#discussion_r56948554>
> :
>
> >  )
> >
> > -type Credentials struct {
> > -	ClientId    string `json:"clientId"`
> > +type Credentials interface {
> > +	fmt.Stringer
> > +	ConfigureAuth(httpRequest *http.Request) error
> > +	getExtField() (string, error)
> > +	HawkCredentials() *hawk.Credentials
>
> FYI, adding a MarshalJSON() ([]byte, error) method to Credentials
> wouldn't be bad, that way you can always get the credentials out again...
>
> I'm not sure we'll need this, as the types should already natively handle
> marshaling (i.e. we do not need to customise the behaviour, we can use the
> system default).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/taskcluster/taskcluster-client-go/pull/13/files/ecd79d67b954c7d0b21dc90c8a2ef95f6273a02e#r56948554>
>